vitrina/orgs/services.py

```diff
@@ -130,63 +322,120 @@ def get_parents(obj: Model) -> list:
+
 def has_perm(
     user: User,  # request.user
     action: Action,
-    obj: (
-        Model  # when action is update, delete
-        | Type[Model]  # when action is create
-    ),
-    # when action is create, object based on which a new objects is created
-    parent: Model | None = None,
+    obj: Model | Type[Model],  # when action is update, delete
+    parent: Model | None = None,  # when action is create, object based on which new object is created
 ) -> bool:
-    if not user.is_authenticated:
-        return False
-
-    if user.is_staff or user.is_superuser:
+    if user.is_authenticated and user.is_superuser:
         return True
-
-    if isinstance(obj, Type):
-        model = obj
-        if parent:
-            nodes = get_parents(parent)
-        else:
-            nodes = []
+    if parent:
+        klass = parent.__class__
+    elif inspect.isclass(obj):
+        klass = obj
+    else:
+        klass = obj.__class__
+    if (
+        action not in EXCLUDED_ACTIONS
+        and klass in DATASET_RELATED_OBJECTS
+        and (dataset := _get_dataset_instance(parent or obj))
+    ):
+        return _has_dataset_perm(user, action, parent or obj, dataset)
     else:
-        model = type(obj)
-        nodes = get_parents(obj)
+        if not user.is_authenticated:
+            return False
 
-    where = []
-    if acl.get((model, action)):
-        for role in acl[(model, action)]:
-            if role == Role.ALL:
-                return True
+        if user.is_staff:
+            return True
+
+        if isinstance(obj, Type):
+            model = obj
+            if parent:
+                nodes = get_parents(parent)
             else:
-                for node in nodes:
-                    if role == Role.AUTHOR:
-                        if is_author(user, node):
-                            return True
-                    elif role == Role.SUPERVISOR:
-                        if is_supervisor(user, node):
-                            return True
-                    else:
-                        ct = ContentType.objects.get_for_model(node)
-                        where.append(
-                            Q(
-                                content_type=ct,
-                                object_id=node.pk,
-                                role=role.value,
+                nodes = []
+        else:
+            model = type(obj)
+            nodes = get_parents(obj)
+
+        where = []
+        if acl.get((model, action)):
+            for role in acl[(model, action)]:
+                if role == Role.AUTHENTICATED:
+                    return True
+                else:
+                    for node in nodes:
+                        if role == Role.AUTHOR:
+                            if is_author(user, node):
+                                return True
+                        elif role == Role.SUPERVISOR:
+                            if is_supervisor(user, node):
+                                return True
```

Comment:
This could be combined as well:

```suggestion
                        if (
                             (role == Role.AUTHOR and is_author(user, node))
                             or (role == Role.SUPERVISOR and is_supervisor(user, node))
                        ):
                            return True
```